layers/+lang/c-c++/README.org

```diff
@@ -52,22 +65,121 @@ by setting the variable =c-c++-default-mode-for-headers= to =c++-mode=.
 #+END_SRC
 
 *Note:* To set the variable for a given project, create a directory local
 variable at the root of your project. More info on directory local variables
 can be found in the [[http://www.gnu.org/software/emacs/manual/html_node/elisp/Directory-Local-Variables.html][dir-locals]].
 
+** Backends
+This layer supports the selection of one of 3 available backends for code navigation etc.
+
+*** RTags
+RTags is a well established clang-based source code indexing tool.
+
+**** External dependencies
+Install the RTags server via [[https://formulae.brew.sh/formula/rtags][homebrew]], the [[https://aur.archlinux.org/packages/rtags/][aur]] or from source according to the instructions [[https://github.com/Andersbakken/rtags][here]].
+N.B. RTags is not supported on Windows at the time of writing, although there is an [[https://github.com/Andersbakken/rtags/issues/770][open issue with some recent activity]] on github.
+
+**** Configuration
+To enable support for =rtags=, set the layer variable
+=c-c++-enable-rtags-support= to =t= in your dotfile.
+
+#+BEGIN_SRC emacs-lisp
+  (setq-default dotspacemacs-configuration-layers
+                '((c-c++ :variables c-c++-enable-rtags-support t)))
+#+END_SRC
+
+This will also enable =company-rtags= to be used as a backend for
+auto-completion (when =auto-completion= layer is included).
+To prevent this, while retaining the rest of Rtags functionality,
+set the variable =c-c++-enable-rtags-support= to ='no-completion=:
+
+#+BEGIN_SRC emacs-lisp
+  (setq-default dotspacemacs-configuration-layers
+                '((c-c++ :variables c-c++-enable-rtags-support 'no-completion)))
+#+END_SRC
+
+*** cquery / ccls (lsp backends)
+[[https://github.com/cquery-project/cquery][cquery]] and [[https://github.com/MaskRay/ccls][ccls]] are alternative implementations of the language server protocol based on libclang. They claim to be more efficient
+than existing tools at indexing large code bases.
+
+**** Features:
+- Cross references (definitions, references, base/derived classes/methods, type instances, ...)
+- Diagnostics
+- Completion with =company-lsp=
+- Semantic highlighting
+- See more on [[https://github.com/cquery-project/cquery/wiki/Emacs]]
+- Cross-platform - functional on Windows, Linux and OSX.
+
+**** External dependencies
+Install one (or both) of the following:
+
+***** cquery server
+Install the =cquery= server. [[https://github.com/cquery-project/cquery/wiki/Getting-started][Instructions]].
+
+***** ccls server
+Install the =ccls= server. [[https://github.com/MaskRay/ccls/wiki/Getting-started][Instructions]].
+
+**** Configuration
+***** Basic
+Enable the [[../../+tools/lsp/README.org][LSP layer]] and select either =cquery= or =ccls= as the backend by adding the following to your dotfile
+
+#+BEGIN_SRC emacs-lisp
+  (setq-default dotspacemacs-configuration-layers
+                '(
+                  lsp
+                  (c-c++ :variables c-c++-backend `cquery) ;or 'ccls
```

Comment:
why do we need to create new layers? we have rtas inside c++ layer, so it's more similar to different backends - rtags, ccls, cquery.
Real issue there is that it's required to enable lsp layers manually. It might be great to use
`(configuration-layer/declare-layer 'base)` but with condition.
@syl20bnr , could you please tell if it's possible?

Reply:
No new layers, just the values of the `c-c++-backend` variable to be prefixed with `lsp`.

Reply:
For me it seem to be unnecessary, what I'd really want is that if we'll be unable to enable automatic dependency from lsp layer, some error message should be printed.
It could be done by addiing empty test package, which will check `c-c++-backend` and if lsp is not available.

Reply:
Re. having to add the `lsp` layer explicitly to your dotfile, creating a `layers.el` with a conditional layer declaration works to get around it:

``` elisp
(when (member c-c++-backend '(lsp-cquery lsp-ccls))
  (configuration-layer/declare-layer 'lsp))
```
is this acceptable @sdwolfz ?

I.e. it appears `config.el` is loaded before `layers.el`.
I'd tried previously using the `spacemacs//c-c++-lsp-backend` function defined in `funcs.el`, but `layers.el` appears to be loaded before it.

Reply:
The current commit amendment still has some logic to display warnings if `lsp` layer not loaded, which should be unnecessary if the conditional layer dependency declaration is acceptable. Easier to strip it out in another amendment if I get the green light than to reinstate it if not :)

Reply:
I think it's good to have it. I don't see it being used in other layers but as long as it does not caus trouble when using a different backend (rtags) then it's a good addition.